src/webui/api/torrentscontroller.cpp

```diff
@@ -1622,3 +1792,246 @@ void TorrentsController::setSSLParametersAction()
 
     torrent->setSSLParameters(sslParams);
 }
+
+void TorrentsController::fetchMetadataAction()
+{
+    requireParams({u"source"_s});
+
+    const QString sourceParam = params()[u"source"_s].trimmed();
+    // must provide some value to parse
+    if (sourceParam.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Must specify URI or hash"));
+
+    const QUrl sourceUrl = QUrl::fromPercentEncoding(sourceParam.toLatin1());
+    const QString source = sourceUrl.toString();
+
+    // metadata has already been requested
+    if (const auto iter = m_torrentSource.constFind(source); iter != m_torrentSource.constEnd())
+    {
+        const BitTorrent::InfoHash infoHash = iter.value();
+        // metadata has already been downloaded
+        if (isMetadataDownloaded(infoHash))
+        {
+            setResult(serializeTorrentInfo(m_torrentMetadata.value(infoHash)));
+        }
+        else
+        {
+            setStatus(APIStatus::Async);
+            setResult(serializeInfoHash(infoHash));
```

Comment:
Can't there be an invalid info hash here due to `m_torrentSource.insert(source, {});` below?

Reply:
Yes, good catch